### PR TITLE
feat(modeling): enhanced poly2 into a complete geometry

### PR DESCRIPTION
**Pull request by z3dev**
These changes enhance the poly2 geometry into a complete geometry with all the common functionality. Hopefully, this will allow additional improvements in the 2D space.

Note: flip was renamed to reverse to align with other 2D geometries.

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

### CODE REVIEW

**Comment by z3dev** [comment]
@platypii please review. i basically ported the old set of changes into this PR for V3.

File: packages/modeling/src/geometries/poly2/isA.test.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import test from 'ava'
+
+import { isA, create } from './index.js'
+
+test('isA: identifies created poly2', (t) => {
+  const p1 = create()
+  const p2 = create([[0, 0, 0], [1, 0, 0], [1, 1, 0]])
```
Comment (platypii) [inline packages/modeling/src/geometries/poly2/isA.test.js:7]: Seems like a poly3 not a poly2

File: packages/modeling/src/geometries/poly2/isConvex.test.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import test from 'ava'
+
+import { isConvex, create } from './index.js'
+
+test('poly3: isConvex() should return correct values', (t) => {
```
Comment (platypii) [inline packages/modeling/src/geometries/poly2/isConvex.test.js:5]: 3 -> 2

File: packages/modeling/src/geometries/poly2/isSimple.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Check whether the given polygon is simple, i.e. does not intersect itself.
+ * @see https://en.wikipedia.org/wiki/Simple_polygon
+ * @param {poly2} polygon - the polygon to interrogate
+ * @returns {Boolean} true if simple
+ * @alias module:modeling/geometries/poly2.isSimple
+ */
+export const isSimple = (polygon) => {
+  const numvertices = polygon.vertices.length
```
Comment (platypii) [inline packages/modeling/src/geometries/poly2/isSimple.js:11]: Nit: `numVertices` instead? camelCase

File: packages/modeling/src/geometries/poly2/isSimple.test.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import test from 'ava'
+
+import { isSimple, create } from './index.js'
+
+test('poly3: isSimple() should return correct values', (t) => {
```
Comment (platypii) [inline packages/modeling/src/geometries/poly2/isSimple.test.js:5]: 3 -> 2

File: packages/modeling/src/geometries/poly2/measureBoundingBox.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import * as vec2 from '../../maths/vec2/index.js'
+
+/**
+ * @param {poly2} polygon - the polygon to measure
+ * @returns {Array} an array of two vectors (2D);  minimum and maximum coordinates
+ * @alias module:modeling/geometries/poly2.measureBoundingBox
+ */
+export const measureBoundingBox = (polygon) => {
+  const vertices = polygon.vertices
+  const numvertices = vertices.length
```
Comment (platypii) [inline packages/modeling/src/geometries/poly2/measureBoundingBox.js:10]: camelCase

**Review by platypii — COMMENTED** [review]
Looks good overall. I have a question of what poly2 should be used for? Should geom2 contain poly2s just like geom3 contains poly3s?

File: packages/modeling/src/geometries/poly2/validate.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    if (vec2.equals(object.vertices[i], object.vertices[(i + 1) % object.vertices.length])) {
+      throw new Error(`poly2 duplicate vertex at ${i}: [${object.vertices[i]}]`)
+    }
+  }
+
+  // check for infinity, nan
+  object.vertices.forEach((vertex) => {
+    if (!vertex.every(Number.isFinite)) {
+      throw new Error(`poly2 invalid vertex ${vertex}`)
+    }
```
Comment (platypii) [inline packages/modeling/src/geometries/poly2/validate.js:41]: Suggestion for checking for vertex dimensionality:
```suggestion
    if (!vertex.every(Number.isFinite)) {
      throw new Error(`poly2 invalid vertex ${vertex}`)
    }
    if (vertex.length != 2) {
      throw new Error(`poly2 invalid vertex ${vertex}`)
    }
```

**Comment by z3dev** [comment]
@platypii thanks for the review. nits are fine. keep those coming...

please review again when you have time.

**Review by platypii — APPROVED** [review]

**Comment by z3dev** [comment]
> Looks good overall. I have a question of what poly2 should be used for? Should geom2 contain poly2s just like geom3 contains poly3s?

Great question... I was thinking that 2D polygons should be a basic geometry like poly3. There are lots of external formats that have this concept. And having something... not really useful... is kind of stupid.

I'm not sure if geom2 should contain a list of poly2, so that's up to you. We will have to see.

Maybe @hrgdavor has some thoughts about 2D polygon support.

**Pull request merged**